src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,829 @@
+                }
+            }
+
+            return modifiedSolution;
+        }
+
+        /// <summary>
+        /// Ties the identifiers within the expression back to their associated parameter.
+        /// </summary>
+        public static async Task<Dictionary<TIdentifierNameSyntax, IParameterSymbol>> MapExpressionToParametersAsync(
+            Document document, TExpressionSyntax expression, CancellationToken cancellationToken)
+        {
+            var nameToParameterDict = new Dictionary<TIdentifierNameSyntax, IParameterSymbol>();
+            var variablesInExpression = expression.DescendantNodes().OfType<TIdentifierNameSyntax>();
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            foreach (var variable in variablesInExpression)
+            {
+                var symbol = semanticModel.GetSymbolInfo(variable, cancellationToken).Symbol;
+                if (symbol is IParameterSymbol parameterSymbol)
+                {
+                    nameToParameterDict.Add(variable, parameterSymbol);
+                }
+            }
+
+            return nameToParameterDict;
+        }
+
+        /// <summary>
+        /// Determines if the expression is something that should have code actions displayed for it.
+        /// Depends upon the identifiers in the expression mapping back to parameters.
+        /// Does not handle params parameters.
+        /// </summary>
+        private static async Task<(bool isParameterized, bool hasOptionalParameter)> ShouldExpressionDisplayCodeActionAsync(
+            Document document, TExpressionSyntax expression, IMethodSymbol methodSymbol,
+            CancellationToken cancellationToken)
+        {
+            var variablesInExpression = expression.DescendantNodes().OfType<TIdentifierNameSyntax>();
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var hasOptionalParameter = false;
+
+            foreach (var parameter in methodSymbol.Parameters)
+            {
+                if (parameter.HasExplicitDefaultValue)
+                {
+                    hasOptionalParameter = true;
+                }
+            }
+
+            foreach (var variable in variablesInExpression)
+            {
+                var parameterSymbol = semanticModel.GetSymbolInfo(variable, cancellationToken).Symbol;
+                if (parameterSymbol is not IParameterSymbol parameter)
+                {
+                    return (false, hasOptionalParameter);
+                }
+                if (parameter.IsParams)
+                {
+                    return (false, hasOptionalParameter);
+                }
+            }
+
+            return (methodSymbol != null && methodSymbol.GetParameters().Any(), hasOptionalParameter);
+        }
```

Comment:
i don't understand the method's purpose, so i can't really review the logic here.

Reply:
I'll message you about this